### PR TITLE
NoJS version of Pinned Post

### PR DESCRIPTION
**Pull request by abeddow91**
## What does this change?
Following from [PR 4132](https://github.com/guardian/dotcom-rendering/pull/4132), this PR introduces a new pinned post component that is written using checkbox input and CSS. This allows the core functionality to work without JS. It gets enhanced with JS through the PinnedPostLiveness component similar to [Liveness](https://github.com/guardian/dotcom-rendering/pull/3869). It feels like the existing Liveness component is a natural home for the pinned post enhancements. However, the existing Liveness component is only for liveblogs and pinned posts feature on both live and dead blogs so it has been separated into a new component. 

## Why?
It allows users without JS enabled to get full use of the component. It also allows for full server side rendering and means the core component doesn't need to live in an island. Also, this pattern is preexisting see https://github.com/guardian/dotcom-rendering/pull/1399 and https://github.com/guardian/frontend/pull/24420 

Adding liveness to the component allows us to add some polish to the component that wouldn't be possible without JS. These include

- [ ] - Scroll on collapse
- [ ] - Tracking 
- [ ] - Keyboard accessibility
- [x] - Remove show more button if the content fits within the container's collapsed max-height

This PR serves to demonstrate how we can add liveness so only those checked above have been implemented in this PR, the remainder are to be completed in separate tickets.

### Before
Screenshot 2022-03-09 at 14 07 48

### After
Screenshot 2022-03-09 at 14 07 38

### CODE REVIEW

File: dotcom-rendering/src/static/icons/pin.svg
Comment (abeddow91) [inline dotcom-rendering/src/static/icons/pin.svg:1]: This will be replaced by a Source Pinned SVG following release

File: common-rendering/src/components/FirstPublished.tsx
Comment (oliverlloyd) [inline common-rendering/src/components/FirstPublished.tsx:15]: What about changing this to `isPinnedPost`? That would feel more declarative and less imperative (I love using those words in PR reviews). Essentially, it's less tell me what to do and more tell me what's happening and let me decide how to react

I'm a little 🤷  here though and you know this code better than me so if you have strong opinions then I'm happy to defer to you

File: common-rendering/src/components/liveBlockContainer.tsx
Comment (oliverlloyd) [inline common-rendering/src/components/liveBlockContainer.tsx:135]: Does it also work if we make the whole line conditional? So we only set a border if not a pinned post. Of course, this idea falls over if the `0` is needed here

File: dotcom-rendering/fixtures/manual/liveBlock.ts
Comment (oliverlloyd) [inline dotcom-rendering/fixtures/manual/liveBlock.ts:1]:
```suggestion
export const liveBlock: Block = {
```

It's always nice if the name of the file matches the name of the export. It makes searching and refactoring easier later on.

File: dotcom-rendering/index.d.ts
Comment (oliverlloyd) [inline dotcom-rendering/index.d.ts:437]: I think you need to run 'make gen-schema' following any edits to this type

File: dotcom-rendering/src/web/components/PinnedPost.stories.tsx
Comment (oliverlloyd) [inline dotcom-rendering/src/web/components/PinnedPost.stories.tsx:83]: 💯

File: dotcom-rendering/src/web/components/PinnedPost.tsx
Comment (oliverlloyd) [inline dotcom-rendering/src/web/components/PinnedPost.tsx:13]: What are these `Props` we're importing here? Normally when a lib exports some props they are explicitly named so you know what you're getting so it's strange to see an export called `Props`. I wonder if this was intentional...?

Comment (oliverlloyd) [reply dotcom-rendering/src/web/components/PinnedPost.tsx:13]: @SiAdcock @sndrs - is Source meant to be exporting this?

File: dotcom-rendering/src/web/components/PinnedPost.tsx
Comment (oliverlloyd) [inline dotcom-rendering/src/web/components/PinnedPost.tsx:129]: I _think_ this can just be `type Props = { ... }`. We're not importing it so there's no need for the export and I don't think we need to extend that other type (but I could be wrong on that)

File: dotcom-rendering/src/web/components/PinnedPost.tsx
Comment (oliverlloyd) [inline dotcom-rendering/src/web/components/PinnedPost.tsx:163]: Would it be possible to use [EditorialButton](https://guardian.github.io/source/?path=/docs/packages-source-react-components-development-kitchen-editorialbutton--playground) from Source here?

It's generally going to be more accessible and it also has the bonus of changing colour based on `theme`.

If you need to add specific attributes I am pretty sure this component extends the standard html button interface so should allow that plus you can also use `cssOverrides` to tweak styles if needed

File: dotcom-rendering/src/web/components/PinnedPostLiveness.importable.tsx
Comment (oliverlloyd) [inline dotcom-rendering/src/web/components/PinnedPostLiveness.importable.tsx:23]: Two things:

1) You can delete the empty object and just have `()`. I don't know why you should do this but some linters require this so my brain has been conditioned to call it out
2) How about `EnhancePinnedPost`? Or something along those lines? Just because 'Liveness' in DCR's context signifies polling for updates but this island isn't really that

File: dotcom-rendering/src/web/components/PinnedPostLiveness.importable.tsx
Comment (oliverlloyd) [inline dotcom-rendering/src/web/components/PinnedPostLiveness.importable.tsx:26]: I think this is fine but it made me wonder if we might ever have an issue where the dom isn't ready but we try to perform operations against it?

In react you get around this using a `useEffect` which only runs after things have rendered. But in our islands world that won't work. In pure js you use the [DOMContentLoaded](https://developer.mozilla.org/en-US/docs/Web/API/Window/DOMContentLoaded_event) event but maybe that's excessive here?

Of course, none of this is relevant to your PR here and is more of a question around how we carry this pattern forward.

File: dotcom-rendering/src/web/lib/LiveBlogRenderer.tsx
Comment (oliverlloyd) [inline dotcom-rendering/src/web/lib/LiveBlogRenderer.tsx:29]: Where you've put the `Island` above does feel like a good location but it could also live here alongside the post itself?

File: dotcom-rendering/src/web/lib/LiveBlogRenderer.tsx
Comment (oliverlloyd) [inline dotcom-rendering/src/web/lib/LiveBlogRenderer.tsx:55]: Should this read `isPinnedPost={false}`?

**Review by oliverlloyd — COMMENTED** [review]
This is really great work. The description of this PR is so good and it really was a pleasure to review this PR, thanks for the effort you clearly put into this.

File: dotcom-rendering/src/web/components/PinnedPostLiveness.importable.tsx
Comment (oliverlloyd) [reply dotcom-rendering/src/web/components/PinnedPostLiveness.importable.tsx:26]: Rereading this comment this is more a note to self for me and you can ignore this

File: dotcom-rendering/src/static/icons/pin.svg
Comment (mxdvl) [inline dotcom-rendering/src/static/icons/pin.svg:3]: Now https://github.com/guardian/source/pull/1292 is merged, can we remove this file?

File: dotcom-rendering/src/web/components/PinnedPost.tsx
Comment (mxdvl) [inline dotcom-rendering/src/web/components/PinnedPost.tsx:20]: Isn’t this already done?

**Review by alinaboghiu — APPROVED** [review]

File: common-rendering/src/components/liveBlockContainer.tsx
Comment (oliverlloyd) [inline common-rendering/src/components/liveBlockContainer.tsx:147]: I think this is reversing the logic here. Does it work if we do this?
```suggestion
						isPinnedPost={isPinnedPost}
```

File: dotcom-rendering/src/web/lib/LiveBlogRenderer.tsx
Comment (oliverlloyd) [inline dotcom-rendering/src/web/lib/LiveBlogRenderer.tsx:100]: This is the non pinned version so I believe it should be false

```suggestion
							isPinnedPost={false}
```

File: dotcom-rendering/fixtures/manual/liveBlock.ts
Comment (abeddow91) [reply dotcom-rendering/fixtures/manual/liveBlock.ts:1]: Absolutely, renamed!

File: dotcom-rendering/src/web/components/PinnedPost.tsx
Comment (abeddow91) [reply dotcom-rendering/src/web/components/PinnedPost.tsx:129]: You're totally right, I've update this.

File: dotcom-rendering/src/web/components/PinnedPost.tsx
Comment (abeddow91) [reply dotcom-rendering/src/web/components/PinnedPost.tsx:13]: Think  this might have been an accident with VSC auto imports. I've removed in favour of local Props

File: dotcom-rendering/src/web/lib/LiveBlogRenderer.tsx
Comment (abeddow91) [reply dotcom-rendering/src/web/lib/LiveBlogRenderer.tsx:100]: Yep you're right - amended.

File: dotcom-rendering/src/web/lib/LiveBlogRenderer.tsx
Comment (abeddow91) [reply dotcom-rendering/src/web/lib/LiveBlogRenderer.tsx:50]: Good call, I've moved the island within the pinnedpost conditional check so its next to the pinned post.

File: dotcom-rendering/src/web/components/PinnedPost.tsx
Comment (oliverlloyd) [inline dotcom-rendering/src/web/components/PinnedPost.tsx:49]: Could we change these rems to pixels?

I realise it's somewhat arbitrary and that there are pros and cons to both approaches but in DCR the convention is pixels so it's really just for consistency.

File: dotcom-rendering/src/web/components/PinnedPost.tsx
Comment (oliverlloyd) [inline dotcom-rendering/src/web/components/PinnedPost.tsx:48]: This will need to use `decidePalette` so that this colour works over different `theme` (Sport blogs for example)

File: dotcom-rendering/src/web/components/PinnedPost.tsx
Comment (oliverlloyd) [inline dotcom-rendering/src/web/components/PinnedPost.tsx:156]: It might be good to have a more specific name for these ids so that we don't get clashes elsewhere (the css names are fine, Emotion handles this for us)

File: common-rendering/src/components/FirstPublished.tsx
Comment (oliverlloyd) [inline common-rendering/src/components/FirstPublished.tsx:48]: Not really sure where this change needs to happen (I'm guessing here but could well be wrong!) but it seems there's too much padding in front of the From time
Screenshot 2022-03-17 at 16 27 04

File: dotcom-rendering/src/web/components/PinnedPost.tsx
Comment (oliverlloyd) [inline dotcom-rendering/src/web/components/PinnedPost.tsx:114]: I'm just dropping this comment here because I don't know where else it should go but do we need a bit more space above the button here?
Screenshot 2022-03-17 at 16 26 26

**Review by oliverlloyd — COMMENTED** [review]
This is really starting to look great! Sorry for all the comments. They're all small things though, so hopefully not too much of a drag and if you have any questions or just want to spread the load I'm very happy to pair on this

File: dotcom-rendering/src/web/components/PinnedPost.tsx
Comment (sndrs) [reply dotcom-rendering/src/web/components/PinnedPost.tsx]: i _think_ that's left in from v3 because people had found it deep in the files in the package and had used it for their own components, iirc @SiAdcock? agree it should have a better name, if it should even be public...

File: common-rendering/src/components/FirstPublished.tsx
Comment (OllysCoding) [inline common-rendering/src/components/FirstPublished.tsx:3]: This speaks to the remaining wider problem than anything in common-rendering is not being linted and is subject to the rules of whoever last edited it.

File: dotcom-rendering/src/web/components/EnhancePinnedPost.importable.tsx
Comment (OllysCoding) [inline dotcom-rendering/src/web/components/EnhancePinnedPost.importable.tsx:11]: Should we be using `undefined` here instead of `null`? cc @mxdvl (I think undefined is generally preferred but I forget why)

Comment (oliverlloyd) [reply dotcom-rendering/src/web/components/EnhancePinnedPost.importable.tsx:11]: I think it needs to be `null` as that is what `querySelector` returns when nothing is found

File: dotcom-rendering/src/web/components/PinnedPost.tsx
Comment (oliverlloyd) [reply dotcom-rendering/src/web/components/PinnedPost.tsx:48]: This is done

File: dotcom-rendering/src/web/components/PinnedPost.tsx
Comment (oliverlloyd) [reply dotcom-rendering/src/web/components/PinnedPost.tsx]: This is done

File: dotcom-rendering/src/web/components/PinnedPost.tsx
Comment (oliverlloyd) [reply dotcom-rendering/src/web/components/PinnedPost.tsx]: This is done

File: common-rendering/src/components/FirstPublished.tsx
Comment (oliverlloyd) [reply common-rendering/src/components/FirstPublished.tsx:48]: This is done

File: dotcom-rendering/src/web/components/PinnedPost.tsx
Comment (oliverlloyd) [reply dotcom-rendering/src/web/components/PinnedPost.tsx:165]: So I think I'm wrong here and that we can't use a classic button component in this situation where we want the css expansion trick

File: common-rendering/src/components/FirstPublished.tsx
Comment (oliverlloyd) [reply common-rendering/src/components/FirstPublished.tsx:55]: It was prettier so it stays!

File: common-rendering/src/components/FirstPublished.tsx
Comment (oliverlloyd) [reply common-rendering/src/components/FirstPublished.tsx:3]: The Guardian configs repo [recommends single quotes](https://github.com/guardian/configs/blob/main/packages/prettier/index.js#L9)

File: dotcom-rendering/src/web/components/EnhancePinnedPost.importable.tsx
Comment (OllysCoding) [reply dotcom-rendering/src/web/components/EnhancePinnedPost.importable.tsx:11]:
```ts
const pinnedPost: Element | undefined = !isServer && window.document.querySelector('[data-gu-marker=pinned-post]') || undefined;
```
Could work?

Comment (oliverlloyd) [reply dotcom-rendering/src/web/components/EnhancePinnedPost.importable.tsx:11]: Why is `undefined` better than `null`? Isn't it that a variable is undefined when it's not been, you know, defined and then in this case one of the possible outcomes is that it becomes `null` which is the value you get when document queries can't find anything.

There used to be a similar thing with `isSignedIn` where `undefined` meant that the check had not completed yet but `false` meant it had completed but the user wasn't signed in. Obviously in this case though the definition is inline so there's never an `undefined` state.

File: dotcom-rendering/src/web/components/PinnedPost.tsx
Comment (oliverlloyd) [reply dotcom-rendering/src/web/components/PinnedPost.tsx:119]: Done

File: dotcom-rendering/src/web/components/EnhancePinnedPost.importable.tsx
Comment (OllysCoding) [reply dotcom-rendering/src/web/components/EnhancePinnedPost.importable.tsx:11]: I don't think it's a massive deal, though like with everything, there are people with stronger opinions who lean towards or away from using one or the other. 

There were some discussions in Client side around it recently and my interpretation had been that 'undefined' was the generally preferred practice. 

I don't see any material difference between the code that's there and the suggestion I wrote besides whether we have a preferred method as a team/codebase, so it's there if we do!

File: common-rendering/src/components/FirstPublished.tsx
Comment (OllysCoding) [reply common-rendering/src/components/FirstPublished.tsx:3]: Mostly concerned that it was wrong in the first place!

File: dotcom-rendering/src/web/components/EnhancePinnedPost.importable.tsx
Comment (oliverlloyd) [reply dotcom-rendering/src/web/components/EnhancePinnedPost.importable.tsx:11]: To be honest, I find these kinds of chained ternaries like `const foo = !isApple && isGreen || canBeEaten` hurts my brains 😵‍💫

Comment (oliverlloyd) [reply dotcom-rendering/src/web/components/EnhancePinnedPost.importable.tsx:11]: Re the null / undefined wars. I'm honestly pretty clueless what it's all about but we do have several examples of this same pattern in the code (I think this code is based on the `Liveness` file) so my preference is to keep things consistent and maybe revisit this separately?

File: common-rendering/src/components/FirstPublished.tsx
Comment (oliverlloyd) [reply common-rendering/src/components/FirstPublished.tsx:3]: Yeah, I see this quote war crop up pretty often and it seems like unnecessary diff noise. I guess the root issue is common not having a config

File: dotcom-rendering/src/web/components/PinnedPost.tsx
Comment (sndrs) [reply dotcom-rendering/src/web/components/PinnedPost.tsx]: deprecated it here https://github.com/guardian/source/pull/1310

File: dotcom-rendering/src/web/components/PinnedPost.tsx
Comment (abeddow91) [reply dotcom-rendering/src/web/components/PinnedPost.tsx]: Yeah, I've been having a look at this this morning and I can't seem to get it working with EditorialButton. If there's a strong desire to use it still, we could ticket this separately so it doesn't block this PR?

Comment (oliverlloyd) [reply dotcom-rendering/src/web/components/PinnedPost.tsx]: No, it's fine. I think that your approach is the right one here 👍

File: dotcom-rendering/cypress/integration/parallel-3/liveblog.interactivity.spec.js
Comment (oliverlloyd) [inline dotcom-rendering/cypress/integration/parallel-3/liveblog.interactivity.spec.js:204]: Using an id here s we can use this element as an anchor tag

File: dotcom-rendering/src/model/json-schema.json
Comment (oliverlloyd) [inline dotcom-rendering/src/model/json-schema.json:398]: This and some other properties crept in here as the schema update routine was not running before

File: dotcom-rendering/src/web/components/Liveness.importable.tsx
Comment (oliverlloyd) [inline dotcom-rendering/src/web/components/Liveness.importable.tsx:29]: I moved these up so they were defined earlier

File: dotcom-rendering/src/web/components/Liveness.importable.tsx
Comment (oliverlloyd) [inline dotcom-rendering/src/web/components/Liveness.importable.tsx:59]: We now insert new blocks immediately after the top-of-blog marker. This marker lives below the pinned post, preventing ordering issues

File: dotcom-rendering/src/web/components/Liveness.importable.tsx
Comment (oliverlloyd) [inline dotcom-rendering/src/web/components/Liveness.importable.tsx:164]: In reviewing this PR I noticed we can improve the way we detect visibility

File: dotcom-rendering/cypress/integration/parallel-3/liveblog.interactivity.spec.js
Comment (jamesgorrie) [reply dotcom-rendering/cypress/integration/parallel-3/liveblog.interactivity.spec.js:204]: Not sure if it makes sense as I am uncertain of what `data-gu-marker` is, but if it's a contract between us and others, maybe we could have both `id` and `data-gu-marker` that serve different purposes?

File: dotcom-rendering/src/model/json-schema.json
Comment (jamesgorrie) [reply dotcom-rendering/src/model/json-schema.json:398]: Should this be automatic?

File: yarn.lock
Comment (jamesgorrie) [inline yarn.lock:20854]: What's the reason for all the `lock` changes?

**Review by jamesgorrie — COMMENTED** [review]
Code looks good, I have a question on the new pattern of adding JS adjacent to the component it's controlling.

File: dotcom-rendering/cypress/integration/parallel-3/liveblog.interactivity.spec.js
Comment (oliverlloyd) [reply dotcom-rendering/cypress/integration/parallel-3/liveblog.interactivity.spec.js:204]: The data attribute is an attempt to move away from using `id`s everywhere but in this case, where we explicitly want an anchor tag, it made more sense to go with that.

I think the data- pattern is emerging. It's nice because you have a more space for description, you can specify both a key and a value, but I don't think there are any hard contracts yet. One contract we _do_ have is to only use classnames for styling. So maybe data attributes can be thought of as alternatives for where classnames used to be used as selectors.

File: yarn.lock
Comment (oliverlloyd) [reply yarn.lock:20854]: I wish I understood this better. All I know is I'm busily going about my business - not making package changes - and then suddenly I run `yarn` and I get this. So I guess either I've merged in `main` and got some weirdness where the lock wasn't updated from another branch or this is somehow date related?

**Review by jamesgorrie — APPROVED** [review]

**Pull request merged**
